### PR TITLE
EZP-30461: Implemented generic Field Type

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30461](https://jira.ez.no/browse/EZP-30461)
| **Bug/Improvement**|no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      |no
| **Tests pass**     | yes
| **Doc needed**     | yes

_This PR is part of DX improvements for eZ Platform 3.0 in terms of implementing Custom Field Types._ 

The goal of this PR is to deliver a base implementation for the 3rd party Custom Field Types but without blocking advanced capabilities. 

https://gist.github.com/adamwojs/3601b0511c8327e5119f3d7df774ed07 is presenting "Hello World" field type based on Generic Field Type. 

## Technical details 

Field Types based on `\eZ\Publish\Core\FieldType\FieldType` already follow the convention that class representing Field Type Value is named "Value" and is in the same namespace that Field Type definition: https://github.com/ezsystems/ezpublish-kernel/blob/7129d217f06d5347d51c51e137ea715d3320414d/eZ/Publish/Core/FieldType/FieldType.php#L446-L454

Generic Field Type apply this convention to several other places: e.g. `\eZ\Publish\SPI\FieldType\FieldType::getEmptyValue` method: https://github.com/ezsystems/ezpublish-kernel/blob/2bcc5c750650f43ec8b0e4c14d56ae372e95db0e/eZ/Publish/Core/FieldType/Generic/Type.php#L41-L46

### Nameable

Generic Field Type implements `\eZ\Publish\SPI\FieldType\Nameable` (required to be able to generate content name when field is part of name-schema or url-schema) by simply converting VO to `string`: 

https://github.com/ezsystems/ezpublish-kernel/blob/2bcc5c750650f43ec8b0e4c14d56ae372e95db0e/eZ/Publish/Core/FieldType/Generic/Type.php#L36-L39

### Value Validation

Generic Field Type use `symfony/validator` component to validate field values. As shown in the example for the simple cases annotations could be used to define constraints. 

In more advanced cases like validation based on the field definition, the `eZ\Publish\Core\FieldType\Generic\Type::getFieldValueConstraints` method has to be overridden. 

https://github.com/ezsystems/ezpublish-kernel/blob/2bcc5c750650f43ec8b0e4c14d56ae372e95db0e/eZ/Publish/Core/FieldType/Generic/Type.php#L112-L115

### Settings Validation 

Generic Field Type is also using `symfony/validator` component to validate field settings. To do so  the `eZ\Publish\Core\FieldType\Generic\Type::getFieldSettingsConstraints` method has to be overridden. For example: 

```php
namespace AppBundle\FieldType\Point2D;

// ... 
use eZ\Publish\Core\FieldType\Generic\Type as GenericType;
use Symfony\Component\Validator\Constraints as Assert;
// ... 

final class Type extends GenericType
{
    protected $settingsSchema = [
        'format' => [
            'default' => "%f"
        ],
    ];

    // ... 

    protected function getFieldSettingsConstraints(): ?Assert\Collection
    {
        return new Assert\Collection([
            'format' => [
                new Assert\NotBlank()
            ]
        ]);
    }
}
```

### Storage Converter 

If `eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter`implementation is not explicitly provided the FT based on Generic Field Type then `eZ\Publish\Core\FieldType\Generic\Converter` will be used as storage converter.  

Internally the field definition / value are serialized to JSON format using `symfony/serializer` component and stored in `data_text`/`data_text5` columns. 

**TODO**:
- [X] Implement feature / fix a bug.
- [X] Implement tests.
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Move Generic Field Type to SPI, drop dependency on Core FieldType (copy what is needed),
- [x] Keep `SymfonySerializerAdapter` in Core as an implementation detail for SPI `ValueSerializer`
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/FieldType/FieldType.php [hunk 1]
```diff
@@ -271,7 +271,7 @@ public function applyDefaultSettings(&$fieldSettings)
      */
     protected function getSortInfo(Value $value)
     {
-        throw new \RuntimeException('Not implemented, yet.');
+        return null;
```
Comment (adamwojs) [inline eZ/Publish/Core/FieldType/FieldType.php]: This change is small but IMHO we should extract it to separate PR because is not only related to Generic Field Type

File: eZ/Publish/Core/Persistence/FieldTypeRegistry.php [hunk 2]
```diff
@@ -91,7 +91,7 @@ public function register($identifier, $fieldType)
      *
      * @return \eZ\Publish\SPI\FieldType\FieldType
      */
-    protected function getCoreFieldType($identifier)
+    public function getCoreFieldType($identifier): FieldTypeInterface
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/FieldTypeRegistry.php]: This change is required for https://github.com/ezsystems/ezpublish-kernel/pull/2632/files#diff-392b90a0512c4db62e37b27822a5bc20R72 but it's definitely not a clean solution. 

Instead of maybe we should introduce dedicated CoreFieldTypeRegistry or expose a method to check the type of FT?

File: eZ/Publish/Core/settings/storage_engines/legacy/field_value_converters.yml [hunk 3]
```diff
@@ -26,14 +26,51 @@ parameters:
...
+        arguments:
+            - [ '@ezpublish.fieldType.generic.serializer.normalizer' ]
+            - [ '@ezpublish.fieldType.generic.serializer.encoder' ]
+
+    ezpublish.fieldType.generic.converter.factory:
+        class: eZ\Publish\Core\FieldType\Generic\ConverterFactory
+        arguments:
+            - '@ezpublish.fieldType.generic.serializer'
+
+#    ezpublish.fieldType.generic.converter:
```
Comment (adamwojs) [inline eZ/Publish/Core/settings/storage_engines/legacy/field_value_converters.yml]: This comment should be removed ;-)

**Comment by emodric** [comment]
Hi. What's the usecase for this field type?

**Comment by adamwojs** [comment]
> Hi. What's the usecase for this field type?

Hi @emodric! I updated the PR description so the purpose of introducing "Generic Field Type" should be clear right now.

**Comment by emodric** [comment]
Thank you @adamwojs !

File: eZ/Publish/Core/FieldType/Generic/Converter.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        $data = $value->dataText;
+        if ($data !== null) {
+            $data = $this->serializer->decode($data, 'json');
+        }
+
+        $fieldValue->data = $data;
+    }
+
+    public function toStorageFieldDefinition(FieldDefinition $fieldDef, StorageFieldDefinition $storageDef): void
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/Converter.php]: I prefer full names of variables instead of shortcuts.$fieldDef -> fieldDefinition and  $storageDef ->  $storageDefinition

File: eZ/Publish/Core/FieldType/Generic/ConverterFactory.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?php
+
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/ConverterFactory.php]: I think you forgot to add copyrights and license

File: eZ/Publish/Core/FieldType/Generic/ConverterProvider.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        $fieldType = $this->fieldTypeRegistry->getCoreFieldType($typeName);
+        if ($this->isGenericFieldType($typeName)) {
+            /** @var \eZ\Publish\Core\FieldType\Generic\Type $fieldType */
+            $converter = $this->converterFactory->createForFieldType(
+                $fieldType->getSettingsClass()
+            );
+
+            // Register and return default converter
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/ConverterProvider.php]: imo a redundant comment , the code is clear

File: eZ/Publish/Core/FieldType/Generic/ConverterProvider.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public function getConverter($typeName): Converter
+    {
+        if ($this->hasConverter($typeName)) {
+            return $this->innerRegistry->getConverter($typeName);
+        }
+
+        $fieldType = $this->fieldTypeRegistry->getCoreFieldType($typeName);
+        if ($this->isGenericFieldType($typeName)) {
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/ConverterProvider.php]: ``if (!$this->isGenericFieldType($typeName)) {
  throw new NotFound($typeName);
}``

File: eZ/Publish/Core/FieldType/Generic/Type.php [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+        return $this->mapConstraintViolationList(
+            $this->validator->validate($fieldSettings, $this->getFieldSettingsConstraints())
+        );
+    }
+
+    /**
+     * @see https://symfony.com/doc/current/validation/raw_values.html
+     *
+     * @return \Symfony\Component\Validator\Constraints\Collection|null
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/Type.php]: redundant phpdoc

File: eZ/Publish/Core/FieldType/Generic/Type.php [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    protected function getFieldSettingsConstraints(): ?Assert\Collection
+    {
+        return null;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\ContentType\FieldDefinition $fieldDefinition
+     *
+     * @return \Symfony\Component\Validator\Constraints\Collection|null
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/Type.php]: redundant phpdoc

File: eZ/Publish/Core/FieldType/Generic/Type.php [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            );
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Returns FQN of class representing Field Type Value.
+     *
+     * @return string
```
Comment (pawbuj) [inline eZ/Publish/Core/FieldType/Generic/Type.php]: redundant phpdoc

File: eZ/Publish/Core/FieldType/Generic/Converter.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+use eZ\Publish\Core\FieldType\FieldSettings;
+use eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter as ConverterInterface;
+use eZ\Publish\Core\Persistence\Legacy\Content\StorageFieldDefinition;
+use eZ\Publish\Core\Persistence\Legacy\Content\StorageFieldValue;
+use eZ\Publish\SPI\Persistence\Content\FieldValue;
+use eZ\Publish\SPI\Persistence\Content\Type\FieldDefinition;
+use Symfony\Component\Serializer\Serializer;
+
+class Converter implements ConverterInterface
```
Comment (bdunogier) [inline eZ/Publish/Core/FieldType/Generic/Converter.php]: How about naming it after what it does (e.g. `JsonSerializerConverter`) instead of "Generic" ?

File: eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/GenericConverter.php [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+use eZ\Publish\Core\FieldType\FieldSettings;
+use eZ\Publish\Core\FieldType\Generic\ValueSerializerInterface;
+use eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter as ConverterInterface;
+use eZ\Publish\Core\Persistence\Legacy\Content\StorageFieldDefinition;
+use eZ\Publish\Core\Persistence\Legacy\Content\StorageFieldValue;
+use eZ\Publish\SPI\Persistence\Content\FieldValue;
+use eZ\Publish\SPI\Persistence\Content\Type\FieldDefinition;
+
+class GenericConverter implements ConverterInterface
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/GenericConverter.php]: final or SPI-extendable?

File: eZ/Publish/Core/settings/storage_engines/legacy/field_value_converters.yml [hunk 13]
```diff
@@ -33,6 +33,11 @@ services:
         class: "%ezpublish.persistence.legacy.field_value_converter.registry.class%"
         lazy: true
 
+    ezpublish.fieldType.ezgeneric.converter:
```
Comment (alongosz) [inline eZ/Publish/Core/settings/storage_engines/legacy/field_value_converters.yml]: Service names should have lowercase letters, though in this case FQCN also applies.

File: eZ/Publish/Core/settings/fieldtypes.yml [hunk 14]
```diff
@@ -601,3 +601,36 @@ services:
...
+        arguments:
+            $normalizer: '@ezpublish.field_type.ezgeneric.value_serializer.symfony.serializer'
+            $denormalizer: '@ezpublish.field_type.ezgeneric.value_serializer.symfony.serializer'
+            $encoder: '@ezpublish.field_type.ezgeneric.value_serializer.symfony.serializer'
+            $decoder: '@ezpublish.field_type.ezgeneric.value_serializer.symfony.serializer'
+
+    eZ\Publish\Core\FieldType\Generic\ValueSerializerInterface:
+        alias: 'eZ\Publish\Core\FieldType\Generic\ValueSerializer\SymfonySerializerAdapter'
+
+    ezpublish.field_type.ezgeneric:
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/settings/fieldtypes.yml]: Actually, this definition is redundant as all dependencies are "autowireable".

File: eZ/Publish/Core/FieldType/Generic/ValueSerializerInterface.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\FieldType\Generic;
+
+interface ValueSerializerInterface
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/FieldType/Generic/ValueSerializerInterface.php]: No. It will be referenced when you override `\eZ\Publish\Core\FieldType\Generic\Type::__construct`  e.g. to pass additional dependencies.

File: eZ/Publish/Core/FieldType/Generic/Type.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\FieldType\Generic;
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/FieldType/Generic/Type.php]: As I said on Slack: Sorry, this is not possible without moving first `\eZ\Publish\Core\FieldType\FieldType` and all related classes to SPI (which is out of story/PR scope).

File: eZ/Publish/Core/settings/fieldtypes.yml [hunk 17]
```diff
@@ -601,3 +601,36 @@ services:
 
     ezpublish.fieldType.ezimageasset:
         alias: eZ\Publish\Core\FieldType\ImageAsset\Type
+
+    ## Generic FT
+
+    ezpublish.field_type.ezgeneric.value_serializer.symfony.normalizer:
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/settings/fieldtypes.yml]: FYI: I decided to simply inject default serializer instead of configuring our own.

**Comment by alongosz** [comment]
Summary of internal sync:
- Move Generic Field Type to SPI, drop dependency on Core FieldType (copy what is needed),
- Keep `SymfonySerializerAdapter` in Core as an implementation detail for `ValueSerializer`
- Make `Generic\Value::__toString()` part of SPI Value contract (can be handled out of scope of this PR),
- Get rid of static checkValueType (can be handled out of scope of this PR).

**Comment by adamwojs** [comment]
PR updated according to @alongosz suggestions (https://github.com/ezsystems/ezpublish-kernel/pull/2632/commits/759158749dae16faa07069e6049983a9a998beb1). Most significant changes:

* Implementation moved to SPI namespace
* `__toString` is now part of the `\eZ\Publish\SPI\FieldType\Value` interface
* `\eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter\SerializableConverter` is meant to be reusable in other field types

File: eZ/Publish/Core/settings/fieldtype_services.yml [hunk 18]
```diff
@@ -54,3 +54,4 @@ services:
             $locationService: '@ezpublish.api.service.location'
             $contentTypeService: '@ezpublish.api.service.content_type'
             $mappings: '$fieldtypes.ezimageasset.mappings$'
+
```
Comment (alongosz) [inline eZ/Publish/Core/settings/fieldtype_services.yml]: Nitpick: seems unnecessary.

File: eZ/Publish/SPI/FieldType/Generic/Type.php [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+    }
+
+    /**
+     * Returns FQN of class representing Field Type Value.
+     *
+     * @return string
+     */
+    protected function getValueClass(): string
+    {
+        return substr_replace(static::class, 'Value', strrpos(static::class, '\\') + 1);
```
Comment (pspanja) [inline eZ/Publish/SPI/FieldType/Generic/Type.php]: FWIW this kind of check for a Value instance was a mistake from a while ago; in case you need to override the Type it also forces you to override the value, which can be a real pain if somewhere in your codebase the original Value is used. If would have been better if each field type had its own implementation instead.

File: eZ/Publish/Core/settings/fieldtype_services.yml (hunk 18 above)
Comment (adamwojs) [reply eZ/Publish/Core/settings/fieldtype_services.yml]: Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/2632/commits/253648a428361200d25718ae008f56d07b1abdc8

File: eZ/Publish/Core/Base/Container/Compiler/GenericFieldTypeConverterPass.php [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+use eZ\Publish\Core\Base\Container\Compiler\Storage\Legacy\FieldValueConverterRegistryPass;
+use eZ\Publish\SPI\FieldType\Generic\Type as GenericType;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\LogicException;
+use Symfony\Component\DependencyInjection\Reference;
+
+final class GenericFieldTypeConverterPass implements CompilerPassInterface
+{
+    public const GENERIC_CONVERTER_SERVICE_ID = 'eZ\Publish\Core\Persistence\Legacy\Content\FieldValue\Converter\SerializableConverter';
```
[the remark this answers is not in the document]
Comment (adamwojs) [reply eZ/Publish/Core/Base/Container/Compiler/GenericFieldTypeConverterPass.php]: Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/2632/commits/253648a428361200d25718ae008f56d07b1abdc8

File: eZ/Publish/SPI/FieldType/Generic/Type.php (hunk 19 above)
Comment (adamwojs) [reply eZ/Publish/SPI/FieldType/Generic/Type.php]: Hi @pspanja! Thank you for your feedback!

> in case you need to override the Type it also forces you to override the value, which can be a real pain if somewhere in your codebase the original Value is used.

This is a common issue with the "convention over configuration" approach. 

> It would have been better if each field type had its own implementation instead.

The use case which you mentioned seems to be very rare. Is it good enough reason for forcing everyone to implement "getValueClass"?

**Comment by adamwojs** [comment]
@mnocon This issue is present in `\eZ\Publish\Core\FieldType\FieldType`  https://github.com/ezsystems/ezpublish-kernel/blob/8cb48719ffadc0745740051c123bb2655477c9a0/eZ/Publish/Core/FieldType/FieldType.php#L351-L354

as well. Roughly looking into related PR (https://github.com/ezsystems/ezpublish-kernel/pull/429) I guess solution will be to remove `$value === null ||` part from expression. WDYT @alongosz?

**Comment by alongosz** [comment]
> I guess solution is to remove `$value === null ||` part from expression. WDYT @alongosz?

Yeah, it doesn't make sense to have this since it's never going to be called like this.

File: eZ/Publish/SPI/FieldType/Generic/Type.php (hunk 19 above)
Comment (emodric) [reply eZ/Publish/SPI/FieldType/Generic/Type.php]: It should, at very least be public, so it can be overriden in a decorator to restore the old value if the type is overriden.

**Comment by adamwojs** [comment]
> Yeah, it doesn't make sense to have this since it's never going to be called like this.

Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/2632/commits/f45823bb2ba0bf345a2543440836b368239951b9. Fix for the `\eZ\Publish\Core\FieldType\FieldType` will be processed as separate PR.

File: eZ/Publish/SPI/FieldType/Generic/Type.php (hunk 19 above)
Comment (alongosz) [reply eZ/Publish/SPI/FieldType/Generic/Type.php]: > This method should, at very least, be public and part of the interface, so it can be overriden in a decorator to restore the old value if the type is overriden.

+1

Comment (adamwojs) [reply eZ/Publish/SPI/FieldType/Generic/Type.php]: Sounds reasonable. Thank you for suggestion @emodric 😉

Comment (alongosz) [reply eZ/Publish/SPI/FieldType/Generic/Type.php]: Note: @adamwojs will provide follow-up with this shortly. It will be easier to review. I'm merging this PR :)

**Comment by adamwojs** [comment]
@bdunogier, @emodric: Your suggestions will be applied as follow up PRs (renaming and making getValueClass public) 😉

**Comment by bdunogier** [comment]
Okay @adamwojs :)
